marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/DefaultSource.java

```diff
@@ -16,28 +16,51 @@
 package com.marklogic.hub.spark.sql.sources.v2;
 
 import com.marklogic.client.ext.helper.LoggingObject;
+import com.marklogic.hub.spark.sql.sources.v2.reader.HubDataSourceReader;
 import com.marklogic.hub.spark.sql.sources.v2.writer.HubDataSourceWriter;
 import org.apache.spark.sql.SaveMode;
 import org.apache.spark.sql.sources.v2.DataSourceOptions;
+import org.apache.spark.sql.sources.v2.ReadSupport;
 import org.apache.spark.sql.sources.v2.StreamWriteSupport;
 import org.apache.spark.sql.sources.v2.WriteSupport;
+import org.apache.spark.sql.sources.v2.reader.DataSourceReader;
 import org.apache.spark.sql.sources.v2.writer.DataSourceWriter;
 import org.apache.spark.sql.sources.v2.writer.streaming.StreamWriter;
 import org.apache.spark.sql.streaming.OutputMode;
 import org.apache.spark.sql.types.StructType;
 
 import java.util.Optional;
 
-public class DefaultSource extends LoggingObject implements WriteSupport, StreamWriteSupport {
+public class DefaultSource extends LoggingObject implements WriteSupport, StreamWriteSupport, ReadSupport {
+
+    private HubDataSourceReader hubDataSourceReader;
+
+    public DefaultSource() {
+        logger.debug("Created: " + toString());
+    }
 
     @Override
     public Optional<DataSourceWriter> createWriter(String writeUUID, StructType schema, SaveMode mode, DataSourceOptions options) {
-        return Optional.of(new HubDataSourceWriter(options.asMap(), schema, false) {
-        });
+        return Optional.of(new HubDataSourceWriter(options.asMap(), schema, false));
     }
 
     @Override
     public StreamWriter createStreamWriter(String queryId, StructType schema, OutputMode mode, DataSourceOptions options) {
         return new HubDataSourceWriter(options.asMap(), schema, true);
     }
+
+    @Override
+    public DataSourceReader createReader(DataSourceOptions options) {
+        // Logging of the HubDataSourceReader's constructor indicates that in a simple Spark test program, this method
+        // is called multiple times. On the first occasion, the getSchema method is called, and then instance is seemingly
+        // discarded. On the second occasion, the planInputPartitions method is called, which allows for partition
+        // readers to then be created. And then this is called on a third occasion for unknown reasons. For performance
+        // reasons then, this class will only create one HubDataSourceReader. Testing has shown that if the Spark
+        // program then calls "format" again on a SQLContext, a new instance of this class - DefaultSource - will be
+        // created, thus ensuring that a new data source reader is created.
+        if (hubDataSourceReader == null) {
```

Comment:
We'll have to keep an eye on this. Creating multiple HubDataSourceReader objects isn't good, but I'm worried there's some downside to caching this instance that I haven't found yet.